### PR TITLE
 Sanitise raw MIDI output by default & add `raw_midi_output` config option

### PR DESCRIPTION
**Pull request by johnnovak**
**TL;DR** 

Yes, this is a real problem. Sometimes, some notes would be missing from the MIDI files captured by Staging, and I got the same results when recording the MIDI data via a loopback driver into a sequencer. It seemed to be totally random, and I've been tearing my hair out for 3 days until I figured out what's going on... I've been doing music and messing around with MIDI for 20+ years now, plus I'm a programmer, and it was far from trivial to even realise what the problem was. So I think it's pretty unlikely that ordinary users who just want to record some MIDI files would be able to identify the problem and fix it themselves. Therefore, I think we should make certain changes that ensure that regular people can record MIDI data from games with confidence and ease. Have you noticed that some DOS MIDI soundtrack captures on YouTube have the following comment? "Re-uploaded again, hopefully no missing notes this time!" Well, I have a hunch it might be related to this issue...

_**Definitely**_ recommend reviewing commit by commit if you value your sanity 😅 (I've done lots of cleanup stuff, all done as separate commits for easy reviewing and bisecting if needed)

# Overview & motivation

 Many MIDI drivers used in games don't fully conform to the MIDI standard. Working with out-of-spec MIDI data recorded from such games in MIDI sequencers can result in various issues, e.g. hanging or missing notes. All Sierra games and the Miles Sound Driver exhibit this problem, while LucasArts games behave correctly.

With this change, we are sanitising all MIDI output by default to fix these issues. Specifically, for "All Notes Off" Channel Mode Messages, and other Channel Mode Messages with an implicit "All Notes Off" behaviour, we are now emitting explicit "Note Off" messages for every active note prior to emitting the Channel Mode message, as mandated by the MIDI specification. This results in no audible difference whatsoever, fixes the missing note problem in certain MIDI sequencers, makes the editing of recorded MIDI files a lot easier, and might even prevent playback problems in some MIDI players.

It is recommended to leave `raw_midi_output` disabled for all games.

For music applications, it _might_ be desirable to enable raw MIDI output in rare cases when one is reliant on some specific non-compliant behaviour of the application (however, I would like to see actual proof of this, but it's safest to introduce the `raw_midi_output` switch to revert to raw output if necessary).

# The problem

This is the MIDI data of the Space Quest V intro music right until the opening scene in REAPER. For those unfamiliar with MIDI sequencers: the horizontal axis is the timeline and the vertical axis is the MIDI note range. Each horizontal bar represents a MIDI note.

I recorded this from real hardware, the MIDI output of an ESS 1868F sound card, specifically. Note this sound card has a perfect MPU-401 implementation, it is _not_ prone to any hanging note bug that plagues many Sound Blaster models, and this issue _**has nothing to do with the Sound Blaster hanging note bug whatsoever**_! I just wanted to make this crystal clear 😄 

I used a hardware recording to illustrate the problem happens on a real machine; the current output recorded using current `main` or any older DOSBox version is identical (using either MIDI capture or recording the actual output via a loopback MIDI driver).

The "All Notes Off" messages are indicated by **red lines** between song/subsong changes. Because the driver does not send Note Off messages for the currently active notes, the **yellow notes** are left hanging when recorded into a MIDI sequencer. That's the first problem: these hanging notes make chopping up a long MIDI recording quite difficult and error prone (i.e. you need to cut the overhanging notes, while taking care not to cut/delete anything else).

But the bigger problem is that this can lead to _missing notes_! The **green notes** are two such places where this happens in this particular recording. The sequence of events leading to a missed note is this:

- Note On message is sent for a given note
- There's a subsong change, and the driver lazily sends the "All Notes Off" event, but _does not_ send Note Off events for the currently playing notes (this violates the spec)
- The note is left hanging when recorded into a sequencer (on playback it's "fine" because the MIDI module turns off the currently playing notes in response to the "All Notes Off" event)
- A bit later, another Note On message is emitted for the same note, followed by a Note Off event
- So now we have _two_ successive Note On events followed by a single Note off event in our recorded MIDI data! In such cases,  the behaviour is sequencer-dependent:
  - some sequencers might not trigger the second Note On at all and might treat it as redundant -- this leads to missing notes, and that's the case in REAPER.
  - Some other program might allow successive Note On messages and retrigger the notes.
 
In any case, the resulting MIDI stream is invalid because Note On messages are not paired up with corresponding Note Offs, as it would be logical, and as it's mandate by the MIDI specification.

image

# The fix

With "MIDI stream sanitisation" enabled, we're simply keeping track of all active notes, and emit Note Off messages when an All Notes Off message is encountered (_prior_ to it, as mandated by the spec). There are other Channel Mode Messages that result in an implicit All Notes Off message; these are handled as well.

This is the result of the MIDI capture with the fix enabled. As you can see, no hanging notes can be observed across All Notes Off boundaries (the vertical lines), which means all Note Ons are properly paired up with Note Offs, and the missing note issue is solved!

image

## Description of `raw_midi_output`

image

# Relevant excerpts from the MIDI Spec

https://archive.org/details/Complete_MIDI_1.0_Detailed_Specification_96-1-3/

Page 20:

> MODE MESSAGES AS ALL NOTES OFF MESSAGES
>
> Messages 123 through 127 also function as All Notes Off messages. They will turn off all voices controlled by the assigned Basic Channel. These messages should not be sent periodically, but only for a specific purpose. **In no case should they be used in lieu of Note-Off commands to turn off notes which have been previously turned on.** Any All Notes Off command (123-127) may be ignored by a receiver with no possibility of notes staying on, since any Note-On command must have a corresponding specific Note-Off command.

Page 24:

> All Notes Off (123) is a mode message which provides an efficient method of turning off all voices turned on via MIDI. While this message is useful for some applications, there is no requirement that a receiver recognize it. **Since recognition of All Notes Off is not required of the receiver, all notes should first be turned off by transmitting individual Note-Off messages prior to sending an All Notes Off.**

Page 25:

> While MIDI devices should be able to respond to the All Notes Off message, **an All Notes Off message should not be sent periodically as part of normal operation. This message should only be used to indicate that the entire MIDI system is "at rest" (i.e. when a sequence has stopped).**

☝🏻 This means using All Notes Off the way some drivers are using (between sub-songs of a single, continuous musical piece) is very questionable in the first place; most likely it's incorrect, and should be considered a hack.

Page A-4:

> ASSIGNMENT OF NOTE ON/OFF COMMANDS
>
> If an instrument receives two or more Note On messages with the same key number and MIDI channel, it must make a determination of how to handle the additional Note Ons. It is up to the receiver as to whether the same voice or another voice will be sounded, or if the messages will be ignored. *The transmitter, however, must send a corresponding Note Off message for every Note On sent.* If the transmitter were to send only one Note Off message, and if the receiver in fact assigned the two Note On messages to different voices, then one note would linger. Since there is no harm or negative side effect in sending redundant Note Off messages this is the recommended practice.

> "ALL NOTES OFF" FUNCTION WHEN SWITCHING MODES
>
> When a receiver is switching between Omni On/Off and Poly or Mono modes, all notes should be turned off . This is to avoid any unexpected behavior when the instrument's mode is switched. Caution should be taken to turn off only those note events received from MIDI and not those played on the receiver's keyboard.

Page A-5:

> If Note Off messages are received while the hold pedal (controller 64 (40H)) is 'on' they must be recognized but not acted upon until the release of the hold pedal. The same is true for the All Notes Off message. A Hold or Sustain pedal 'On' message takes priority over Note Off and All Notes Off until it is released.

### CODE REVIEW

**Comment by johnnovak** [comment]
I'll hold off fixing the sanitiser issues until the PR comments arrive.

File: src/hardware/ne2000.cpp [hunk 1]
```diff
@@ -1533,15 +1533,19 @@ void NE2K_ShutDown(Section* sec) {
...
-	sec->AddDestroyFunction(&NE2K_ShutDown,true);
-	if(!test->load_success) {
+
+	constexpr auto changeable_at_runtime = true;
+	sec->AddDestroyFunction(&NE2K_ShutDown, changeable_at_runtime);
+
+	if (!test->load_success) {
 		delete test;
-		test=0;
+		test = 0;
```
Comment (kcgen) [inline src/hardware/ne2000.cpp:1548]: Maybe ` = nullptr`

File: src/midi/midi.cpp [hunk 2]
```diff
@@ -54,70 +53,73 @@ uint8_t MIDI_evt_len[256] = {
...
-  3,3,3,3, 3,3,3,3, 3,3,3,3, 3,3,3,3,  // 0xe0
+  3,3,3,3, 3,3,3,3, 3,3,3,3, 3,3,3,3,  // 0xe0 -- Pitch Bend
 
-  0,2,3,2, 0,0,1,0, 1,0,1,1, 1,0,1,0   // 0xf0
+  0,2,3,2, 0,0,1,0, 1,0,1,1, 1,0,1,0   // 0xf0 -- System Exclusive
 };
 // clang-format on
 
-MidiHandler* handler_list = 0;
+static MidiHandler* handler_list = 0;
```
Comment (kcgen) [inline src/midi/midi.cpp:71]: ` = nullptr`.

File: src/midi/midi.cpp [hunk 3]
```diff
@@ -168,6 +217,85 @@ uint8_t get_midi_channel(const uint8_t channel_status)
...
+// dealing with recorded MIDI data. This can lead lead to missing notes, and
+// it makes editing long MIDI recordings containing multiple songs very
+// difficult and error-prone.
+//
+// See page 20, 24, 25 and A-4 of the "The Complete MIDI 1.0 Detailed
+// Specification" document version 96.1, third edition (1996, MIDI
+// Manufacturers Association) for further details
+//
+// https://archive.org/details/Complete_MIDI_1.0_Detailed_Specification_96-1-3/
+//
```
Comment (kcgen) [inline src/midi/midi.cpp:269]: Extremely helpful discussion of the issue and the use of the new sanitize feature!

File: src/hardware/mpu401.cpp [hunk 4]
```diff
@@ -150,16 +158,16 @@ static void MPU401_WriteCommand(io_port_t, const io_val_t value, io_width_t)
...
 		case 2:
-			MIDI_RawOutByte(0xfa);
+			MIDI_RawOutByte(MidiStatus::Start);
 			mpu.clock.cth_counter = mpu.clock.cth_savecount = 0;
 			break;
 		case 3:
-			MIDI_RawOutByte(0xfb);
+			MIDI_RawOutByte(MidiStatus::Continue);
 			mpu.clock.cth_counter = mpu.clock.cth_savecount;
 			break;
```
Comment (kcgen) [inline src/hardware/mpu401.cpp:172]: All of a sudden, the switch statement became self-documenting!

File: src/midi/midi_alsa.cpp [hunk 5]
```diff
@@ -45,7 +45,7 @@ static void for_each_alsa_seq_port(port_action_t action)
 	// be called before that sequencer is created.
 	snd_seq_t *seq = nullptr;
 	if (snd_seq_open(&seq, "default", SND_SEQ_OPEN_OUTPUT, 0) != 0) {
-		LOG_MSG("ALSA: Error: Can't open MIDI sequencer");
+		LOG_WARNING("MIDI:ALSA: Can't open MIDI sequencer");
```
Comment (kcgen) [inline src/midi/midi_alsa.cpp:48]: `MIDI:ALSA:` -> nice clarification! Especially useful because there is a big difference in how ALSA handles MIDI stream versus PCM (and diagnosing them is quite a different task.. with different kernel modules underpinning each).

File: src/midi/midi_win32.h [hunk 6]
```diff
@@ -106,7 +106,7 @@ class MidiHandler_win32 final : public MidiHandler {
 	void PlaySysex(uint8_t *sysex, size_t len) override
 	{
 		if (WaitForSingleObject (m_event, 2000) == WAIT_TIMEOUT) {
-			LOG(LOG_MISC,LOG_ERROR)("Can't send midi message");
+			LOG_WARNING("MIDI:WIN32: Can't send midi message.");
```
Comment (kcgen) [inline src/midi/midi_win32.h:109]: (maybe drop the trailing period?)

File: src/cpu/cpu.cpp [hunk 7]
```diff
@@ -2435,9 +2435,15 @@ void CPU_ShutDown([[maybe_unused]] Section* sec) {
...
+void CPU_Init(Section* sec)
+{
+	assert(sec);
+
 	test = new (std::nothrow) CPU(sec);
-	sec->AddDestroyFunction(&CPU_ShutDown,true);
+
+	constexpr auto changeable_at_runtime = true;
+	sec->AddDestroyFunction(&CPU_ShutDown, changeable_at_runtime);
 }
```
Comment (kcgen) [inline src/cpu/cpu.cpp:2446]: For anyone reading the code and investigating some issue, seeing literal bools sprinkled about the code, like:

 - `AddFunction(func, true)`
 - `AddFunction(func, false)`

You either device to to read past them with the uneasy feeling of, *"well, I guess I don't know what happening with these functions ..."*, or you break your concentration and go lookupt hose calls just to double check (and yeah.. I can't remember how many times I had to grep and look them up!)

All of that to say, awesome full-pass cleanup commit, @johnnovak!

Comment (kcgen) [reply src/cpu/cpu.cpp:2446]: One left in `src/dosbox.cpp` :-)

git grep 'Add.*Function' | grep 'true)\|false)'

```c++
secprop->AddInitFunction(&DOS_KeyboardLayout_Init, true);
```

File: src/midi/midi.cpp (hunk 3 above)
Comment (johnnovak) [reply src/midi/midi.cpp:269]: Was afraid if it might be a bit too much info for some, but then I started thinking it would do the project a lot of good if literally 50% of the total LOC was helpful documentation like this, explaining varius bits and pieces about the hardware, standards, etc. Glad you appreciate it! 🎉

File: src/hardware/mpu401.cpp (hunk 4 above)
Comment (johnnovak) [reply src/hardware/mpu401.cpp:172]: Yeah, that's exactly what I'm striving for! 😄

File: src/midi/midi_win32.h (hunk 6 above)
Comment (johnnovak) [reply src/midi/midi_win32.h:109]: Sure.

File: src/cpu/cpu.cpp (hunk 7 above)
Comment (johnnovak) [reply src/cpu/cpu.cpp:2446]: Ah, one slipped through somehow! 😄

Thanks for flagging it. I think we'll need to do this pattern for a long time, or maybe forever, because it doesn't look like named arguments are coming to C++ anytime soon (or ever, because it's hard to implement them in a backward compatible manner).

File: src/midi/midi.cpp [hunk 8]
```diff
@@ -580,6 +713,16 @@ void init_midi_dosbox_settings(Section_prop& secprop)
...
+
+	auto* bool_prop = secprop.Add_bool("raw_midi_output", when_idle, false);
+	assert(bool_prop);
+	bool_prop->Set_help(
+	        "Enable raw, unaltered MIDI output (disabled by default).\n"
+	        "Many MIDI drivers used in games don't fully conform to the MIDI standard.\n"
+	        "Working with out-of-spec MIDI data recorded from such games in MIDI\n"
+	        "sequencers can result in various issues, e.g. hanging or missing notes.\n"
+	        "DOSBox tries to fix such problems by default; this should not normally affect\n"
+	        "the output in any audible way.");
```
[the remark this answers is not in the document]
Comment (johnnovak) [reply src/midi/midi.cpp:725]: Yeah, and I'm not 100% happy with the wording either. Spent a long time on it... writing is hard 😅 (I can attest that by spending many dozens of hours on the doco project already...)

I think I want to keep the config name `raw_midi_output` though. Did not want to make it too specific and long (e.g., `explicit_note_offs_on_channel_mode_change_midi_messages` 😮‍💨, `compliant_midi_mode` -- compliant with what? 🤦🏻  ), and wanted to prevernt bizarre names too (e.g., `sanitise_midi_output`.

[the remark this answers is not in the document]
Comment (kcgen) [reply src/midi/midi.cpp:725]: Oh, yes - the conf name (raw_mini_output) is perfect and is very self-documenting as-is. It tells me that there is some checking and potential rectifying being done in the default scenario. 

So I think the description just has to basically confirm that, and to let users know that if midi doesn't sound right, they could try letting the raw steam out (because maybe some "defect" is actually part of the original experience, or some like that 😅!)

**Review by kcgen — APPROVED** [review]
Feel free to merge when you like, @johnnovak.
This is a huge clean up, all the future readers of the MIDI code thank you! 🎹 🚀

**Comment by johnnovak** [comment]
@kcgen All green now and I've done some further improvements. Holding off the merge until you get a chance to have a look at them.

**Comment by kcgen** [comment]
Getting a stack-underflow in Eric the Unready using MT-32.

It's because the prior code used (and assumed) fixed-length message arrays passed to the `PlayMsg(uint8_t*)` functions.

```diff
diff --git a/src/midi/midi_fluidsynth.cpp b/src/midi/midi_fluidsynth.cpp
index c97523ab7..340a588ca 100644
--- a/src/midi/midi_fluidsynth.cpp
+++ b/src/midi/midi_fluidsynth.cpp
@@ -602,7 +602,10 @@ uint16_t MidiHandlerFluidsynth::GetNumPendingAudioFrames()
 // The request to play the channel message is placed in the MIDI work FIFO
 void MidiHandlerFluidsynth::PlayMsg(const uint8_t* msg)
 {
-       std::vector<uint8_t> message(msg, msg + MaxMidiMessageLen);
+       const auto status_byte = *msg;
+       const auto msg_len = MIDI_message_len_by_status[status_byte];
+
+       std::vector<uint8_t> message(msg, msg + msg_len);
        MidiWork work{std::move(message),
                      GetNumPendingAudioFrames(),
                      MessageType::Channel};
diff --git a/src/midi/midi_mt32.cpp b/src/midi/midi_mt32.cpp
index 0adffcccf..31cdb8105 100644
--- a/src/midi/midi_mt32.cpp
+++ b/src/midi/midi_mt32.cpp
@@ -756,8 +756,10 @@ uint16_t MidiHandler_mt32::GetNumPendingAudioFrames()
 // The request to play the channel message is placed in the MIDI work FIFO
 void MidiHandler_mt32::PlayMsg(const uint8_t *msg)
 {
-       constexpr auto channel_len = 4;
-       std::vector<uint8_t> message(msg, msg + channel_len);
+       const auto status_byte = *msg;
+       const auto msg_len = MIDI_message_len_by_status[status_byte];
+
+       std::vector<uint8_t> message(msg, msg + msg_len);
        MidiWork work{std::move(message),
                      GetNumPendingAudioFrames(),
                      MessageType::Channel};
```

But looking at all the PlayMsg() handlers, there are some that assume up to three bytes:

```c++
void MidiHandler_alsa::PlayMsg(const uint8_t* msg)
{
	const auto status_byte = msg[0];

	ev.type            = SND_SEQ_EVENT_OSS;
	ev.data.raw32.d[0] = status_byte;
	ev.data.raw32.d[1] = msg[1];
	ev.data.raw32.d[2] = msg[2];
```

Apple's coreaudio:

```c++
void PlayMsg(const uint8_t *msg) override
{
	MusicDeviceMIDIEvent(m_synth, msg[0], msg[1], msg[2], 0);
```

Windows API seems to expect a full 4-byte packed integer, and the code also assumes 3-bytes are always available:

```c++
void PlayMsg(const uint8_t* data) override
{
	const auto status  = data[0];
	const auto data1   = data[1];
	const auto data2   = data[2];
	const uint32_t msg = status + (data1 << 8) + (data2 << 16);

	midiOutShortMsg(m_out, msg);
```


(for ref, here's the backtrack)

```text
2023-03-04 07:18:33.591 | MT32: Initialized CM-32L/LAPC-I Control v1.02 from /usr/share/mt32-rom-data

2023-03-04 07:18:33.594 | MT32: Operating at 48000 Hz without resampling
[New Thread 0x7fffd63366c0 (LWP 770416)]

2023-03-04 07:18:33.594 | MIDI: Opened device: mt32

stack-buffer-overflow on address 0x7ffff458cfe3 at pc 0x5555569ec872 bp 0x7fffffffc680 sp 0x7fffffffbe48
READ of size 4 at 0x7ffff458cfe3 thread T0


message = std::vector of length -105690558186800, capacity 4 = {190 '\276', 190 '\276', 190 '\276', 190 '\276', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000', 0 '\000'...}
        work = {message = std::vector of length -3, capacity 106377750001333 = {<error reading variable: Cannot access memory at address 0x3>
#24 0x0000555558181ccd in MidiHandler::HaltSequence() (this=0x55555badc060 <mt32_instance>) at ../../src/midi/midi.cpp:430
        channel = 0 '\000'
        message = "\260{"
#25 0x000055555818233e in MIDI_HaltSequence() () at ../../src/midi/midi.cpp:450
#26 0x0000555557d5efe5 in MPU401_Reset() () at ../../src/hardware/mpu401.cpp:706
#27 0x0000555557d6450c in MPU401::MPU401(Section*) (this=0x60600000f020, configuration=0x61300000cec0) at ../../src/hardware/mpu401.cpp:783
        section = 0x61300000cec0
Python Exception <class 'gdb.error'>: There is no member named _M_p.
        mpu_type = 
        port_0x330 = 816
        port_0x331 = 817
#28 0x0000555557d57288 in MPU401_Init(Section*) (sec=0x61300000cec0) at ../../src/hardware/mpu401.cpp:807
        changeable_at_runtime = false
#29 0x0000555556ca54cc in Section::ExecuteInit(bool) (this=0x61300000cec0, initall=true) at ../../src/misc/setup.cpp:1023
        fn = @0x61500001a720: {function = 0x555557d571e0 <MPU401_Init(Section*)>, changeable_at_runtime = true}

                    __range1 = std::deque with 3 elements = {{function = 0x5555581836d0 <MIDI_Init(Section*)>, changeable_at_runtime = true}, {function = 0x55555817e9c0 <init_midi_state(Section*)>, changeable_at_runtime = true}, {function = 0x555557d571e0 <MPU401_Init(Section*)>, changeable_at_runtime = true}}
        __begin1 = {function = 0x555557d571e0 <MPU401_Init(Section*)>, changeable_at_runtime = true}
        __end1 = {function = 0xbebebebebebebebe, changeable_at_runtime = 190}
#30 0x0000555556ca4a27 in Config::Init() const (this=0x616000000380) at ../../src/misc/setup.cpp:994
```

In the previous code,

**Comment by kcgen** [comment]
More progress, libmt32emu's `playMsg(uint32_t)` also takes in a little-endian 32-bit int integer:

## MT-32

```c++
void MidiHandler_mt32::ProcessWorkFromFifo() {
	...
	const auto msg_words = reinterpret_cast<const uint32_t*>(work.message.data());
	service->playMsg(SDL_SwapLE32(*msg_words));
	...
```

Hopefully can re-use the same fix you put in for Win32.

**Comment by johnnovak** [comment]
Thanks for these and the testing @kcgen. I've tested it with FluidSynth on the Mac only, but forgot about all that buffer trickery because I did that refactor first a while ago...

**Comment by johnnovak** [comment]
Phew, that was a long ride @kcgen 😅 

I've fixed things up and have successfully tested the following:

- macOS: CoreMidi, MT-32 & FluidSynth (using the sanitizer build)
- Windows: Win32 MIDI, MT-32, FluidSynth (release build)

If you could give it a quick test on Linux using ALSA that would be great, otherwise I'll just merge.

**Comment by kcgen** [comment]
Nice one, @johnnovak -- UBSAN all passing here! :rocket:
